src/common/eeprom.cpp

```diff
@@ -509,3 +595,132 @@ int8_t eeprom_test_PUT(const unsigned int bytes) {
     }
     return res_flag;
 }
+
+///////////////////////////////////////////////////////////////////////////////
+/// Sheets profile methods
+uint32_t sheet_next_calibrated() {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    uint8_t index = variant_get_ui8(eeprom_get_var(EEVAR_ACTUAL_SHEET));
+
+    for (int8_t i = 1; i < MAX_SHEETS; ++i) {
+        if (sheet_select((index + i) % MAX_SHEETS))
+            return (index + i) % MAX_SHEETS;
+    }
+#endif
+    return 0;
+}
+
+bool sheet_is_calibrated(uint32_t index) {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    uint16_t profile_address = eeprom_var_addr(EEVAR_SHEET_PROFILE0 + index);
+    float z_offset = FLT_MAX;
+    st25dv64k_user_read_bytes(profile_address + MAX_SHEET_NAME_LENGTH,
+        &z_offset, sizeof(float));
+    return !nearlyEqual(z_offset, FLT_MAX, 0.001f);
+#else
+    return index == 0;
+#endif
+}
+
+bool sheet_select(uint32_t index) {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    if (index >= MAX_SHEETS || !sheet_is_calibrated(index))
+        return false;
+    uint16_t actual_sheet_address = eeprom_var_addr(EEVAR_ACTUAL_SHEET);
+    st25dv64k_user_write(actual_sheet_address, index);
+    return true;
+#else
+    return index == 0;
+#endif
+}
+
+bool sheet_calibrate(uint32_t index) {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    if (index >= MAX_SHEETS)
+        return false;
+    uint16_t actual_sheet_address = eeprom_var_addr(EEVAR_ACTUAL_SHEET);
+    st25dv64k_user_write(actual_sheet_address, index);
+    return true;
+#else
+    return index == 0;
+#endif
+}
+
+bool sheet_reset(uint32_t index) {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    if (index >= MAX_SHEETS)
+        return false;
+    uint8_t actual = variant_get_ui8(eeprom_get_var(EEVAR_ACTUAL_SHEET));
+    uint16_t profile_address = eeprom_var_addr(EEVAR_SHEET_PROFILE0 + index);
+    float z_offset = FLT_MAX;
+    st25dv64k_user_write_bytes(profile_address + MAX_SHEET_NAME_LENGTH,
+        &z_offset, sizeof(float));
+    if (actual == index)
+        sheet_next_calibrated();
+    return true;
+#else
+    return false;
+#endif
+}
+
+uint32_t sheet_number_of_calibrated() {
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    uint32_t count = 1;
+    for (int8_t i = 1; i < MAX_SHEETS; ++i) {
+        if (sheet_is_calibrated(i))
+            ++count;
+    }
+    return count;
+#else
+    return 1;
+#endif
+}
+
+uint32_t sheet_current_name(char *buffer, uint32_t length) {
+    if (!buffer || !length)
+        return 0;
+#if (EEPROM_FEATURES & EEPROM_FEATURE_SHEETS)
+    uint8_t index = variant_get_ui8(eeprom_get_var(EEVAR_ACTUAL_SHEET));
+    return sheet_name(index, buffer, length);
+#else
+    memcpy(buffer, "DEFAULT", MAX_SHEET_NAME_LENGTH - 1);
```

Comment:
having "DEFAULT" as a static const char def[] = "DEFAULT" may force the compiler to skip loading the string into RAM. Gcc7 is may be smart enough to optimize it, but on the AVR this was eating RAM

Reply:
I did not know that name should I return in case that the feature sheets_profiles will be disabled by compilation, but I wanted to keep the same approach as in every other sheet profile function.

Reply:
ok